# **Kafka Data Stream Processing with Python**
## **Kafka Connection & Creating Random Measurements**

## Kafka Python Library Installation

In [10]:
#%pip install kafka-python

## The Used Packages

In [11]:
from kafka import KafkaProducer
import json
import time
import random

## The Message Broker Connection Test

In [12]:
bootstrap_servers = 'localhost:9092'

producer = KafkaProducer(bootstrap_servers=bootstrap_servers)

topic = 'measurements'

message = b"Message broker connection test!"

producer.send(topic, message)

producer.close()

### The connection has been checked via the terminal as it is shown below.

**The command written on the terminal:** C:\kafka\bin\windows>kafka-console-consumer.bat --topic measurements --bootstrap-server localhost:9092 --from-beginning

**The output on the terminal:**

Message broker connection test!

## Creating Random Measurements('temperature', 'moisture', 'wind_speed') Periodicaly
#### Note: The sending is limited to a specific time duration (90 seconds).

In [13]:
def measurement_generator():
    measurement = {
        'measurement_type': random.choice(['temperature', 'moisture', 'wind_speed']),
        'value': round(random.uniform(0, 100), 2)
    }
    return measurement

def measurement_sender(producer, topic, duration):
    start_time = time.time()
    while time.time() - start_time < duration:
        measurement_data = measurement_generator()
        
        producer.send(topic, value=json.dumps(measurement_data).encode('utf-8'))
        print(f"Sent: {measurement_data}")

        time.sleep(5)

def perform_measurements():
    bootstrap_servers = 'localhost:9092'
    
    topic = 'measurements'
    
    producer = KafkaProducer(bootstrap_servers=bootstrap_servers)
    
    # the sending is limited to a specific time duration (90 seconds).
    duration = 90
    measurement_sender(producer, topic, duration)
    
    producer.close()

if __name__ == "__main__":
    perform_measurements()


Sent: {'measurement_type': 'temperature', 'value': 56.16}
Sent: {'measurement_type': 'temperature', 'value': 60.96}
Sent: {'measurement_type': 'temperature', 'value': 36.95}
Sent: {'measurement_type': 'wind_speed', 'value': 57.84}
Sent: {'measurement_type': 'moisture', 'value': 28.21}
Sent: {'measurement_type': 'wind_speed', 'value': 99.95}
Sent: {'measurement_type': 'moisture', 'value': 55.56}
Sent: {'measurement_type': 'wind_speed', 'value': 4.85}
Sent: {'measurement_type': 'wind_speed', 'value': 2.71}
Sent: {'measurement_type': 'moisture', 'value': 47.39}
Sent: {'measurement_type': 'temperature', 'value': 75.43}
Sent: {'measurement_type': 'temperature', 'value': 64.34}
Sent: {'measurement_type': 'temperature', 'value': 5.4}
Sent: {'measurement_type': 'temperature', 'value': 36.37}
Sent: {'measurement_type': 'wind_speed', 'value': 50.78}
Sent: {'measurement_type': 'moisture', 'value': 16.74}
Sent: {'measurement_type': 'temperature', 'value': 12.4}
Sent: {'measurement_type': 'wind_spe

### Sending the Random Measurements ('temperature', 'moisture', 'wind_speed') messages has been also checked via the terminal as it is shown below.

**The command written on the terminal:** C:\kafka\bin\windows>kafka-console-consumer.bat --topic measurements --bootstrap-server localhost:9092 --from-beginning

**The output on the terminal:**

{"measurement_type": "temperature", "value": 56.16}

{"measurement_type": "temperature", "value": 60.96}

{"measurement_type": "temperature", "value": 36.95}

{"measurement_type": "wind_speed", "value": 57.84}

{"measurement_type": "moisture", "value": 28.21}

{"measurement_type": "wind_speed", "value": 99.95}

{"measurement_type": "moisture", "value": 55.56}

{"measurement_type": "wind_speed", "value": 4.85}

{"measurement_type": "wind_speed", "value": 2.71}

{"measurement_type": "moisture", "value": 47.39}

{"measurement_type": "temperature", "value": 75.43}

{"measurement_type": "temperature", "value": 64.34}

{"measurement_type": "temperature", "value": 5.4}

{"measurement_type": "temperature", "value": 36.37}

{"measurement_type": "wind_speed", "value": 50.78}

{"measurement_type": "moisture", "value": 16.74}

{"measurement_type": "temperature", "value": 12.4}

{"measurement_type": "wind_speed", "value": 87.5}